src/services/jsTyping.ts

```diff
@@ -5,49 +5,190 @@
+            if(inferredTypings[typing]) {
+                cachedTypingPaths.push(inferredTypings[typing]);
+            }
+            else {
+                newTypingNames.push(typing);
+            }
+        }
         
-        return { cachedTypingPaths, newTypings: newTypingNames };
+        return { cachedTypingPaths, newTypingNames };
+    }
+    
+    function mergeTypings(typingNames: string[]) {
+        for (let typing of typingNames) {
+            if (!inferredTypings.hasOwnProperty(typing)) {
+                inferredTypings[typing] = undefined;
+            }
+        }
+    }
+
+    function getTypingNamesFromPackageJson(packageJsonPath: string) {
+        if (_host.fileExists(packageJsonPath)) {
+            // Guide against malformed package.json
+            try{
+                let packageJsonDict = JSON.parse(_host.readFile(packageJsonPath));
+                if (packageJsonDict.hasOwnProperty("dependencies")) {
+                    mergeTypings(Object.keys(packageJsonDict.dependencies));
+                }
+            }
+            catch(e) {
+            }
+        }
+    }
+
+    function getTypingNamesFromBowerJson(bowerJsonPath: string) {
+        if (_host.fileExists(bowerJsonPath)) {
+            // Guide against malformed package.json
+            try{
+                let packageJsonDict = JSON.parse(_host.readFile(bowerJsonPath));
+                if (packageJsonDict.hasOwnProperty("dependencies")) {
+                    mergeTypings(Object.keys(packageJsonDict.dependencies));
+                }
+            }
+            catch(e) {
+            }
+        }
+    }
+
+    /**
+     * Infer typing names from given file names. For example, the file name "jquery-min.2.3.4.js"
+     * should be inferred to the 'jquery' typing name; and "angular-route.1.2.3.js" should be inferred
+     * to the 'angular-route' typing name.
+     * @param fileNames are the names for source files in the project
+     * @param safeList is the list of names that we are confident they are library names that requires typing
+     */
+    function getTypingNamesFromSourceFileNames(fileNames: string[], safeList: string[]) {
+        safeList = safeList.map(s => s.toLowerCase());
+        fileNames = fileNames.map(f => f.toLowerCase());
+        let exactlyMatched: string[] = []
+        let notExactlyMatched: string[] = []
+        for (let fileName of fileNames) {
+            let baseName = ts.getBaseFileName(fileName);
+            let baseNameWithoutExtension = baseName.substring(0, baseName.lastIndexOf("."));
+            (safeList.indexOf(baseNameWithoutExtension) >= 0) ? exactlyMatched.push(baseNameWithoutExtension) : notExactlyMatched.push(baseNameWithoutExtension);
+        }
+        
+        let regex = /((?:\.|-)min(?=\.|$))|((?:-|\.)\d+)/g;
+        notExactlyMatched = notExactlyMatched.map(f => f.replace(regex, ""));
+        let typingNames = exactlyMatched.concat(notExactlyMatched.filter(f => ts.contains(safeList, f)));

```

Comment:
Meant to ask this in my CR - is ts.filter() preferable to using .filter() - optimized?